connect.go

```diff
@@ -1,257 +1,190 @@
+		client.SLogger.Error("keepalive list tags failed", slog.Any("err", err))
+		return
 	}
 
-	if fwopenresphdr.Status != 0x00 {
-		client.Logger.Printf("bad status on large forward open header. got %x. Falling back to small forward open", fwopenresphdr.Status)
-		client.ConnectionSize = 502
-
-		// we have to do something different for small connection sizes.
-		fwd_open, err := client.newForwardOpenStandard()
-		if err != nil {
-			return fmt.Errorf("couldn't create forward open. %w", err)
-		}
-		s := binary.Size(fwd_open)
-		_ = s
-		items0 := make([]CIPItem, 2)
-		items0[0] = CIPItem{Header: cipItemHeader{ID: cipItem_Null}}
-		items0[1] = fwd_open
-		itemdata, err := serializeItems(items0)
-		if err != nil {
-			return err
-		}
-		hdr, dat, err := client.send_recv_data(cipCommandSendRRData, itemdata)
-		if err != nil {
-			return err
-		}
-		_ = hdr
-		if hdr.Status != 0x00 {
-			return fmt.Errorf("small Forward Open Failed. code %v", hdr.Status)
-		}
-		// header before items
-		preitem = msgPreItemData{}
-		err = binary.Read(dat, binary.LittleEndian, &preitem)
-		if err != nil {
-			return fmt.Errorf("problem reading items header from forward open req. %w", err)
-		}
-
-		items, err = readItems(dat)
-		if err != nil {
-			return fmt.Errorf("problem reading items from forward open req. %w", err)
-		}
+	t := time.NewTicker(client.KeepAliveFrequency)
+	defer t.Stop()
+	for {
+		select {
+		case <-t.C:
+			if !client.Connected {
+				client.SLogger.Warn("keepalive failed. not connected")
+				return
+			}
 
-		fwopenresphdr = msgCIPMessageRouterResponse{}
-		err = items[1].DeSerialize(&fwopenresphdr)
-		if err != nil {
-			return fmt.Errorf("error deserializing forward open response header. %w", err)
-		}
-		extended_status = make([]byte, fwopenresphdr.Status_Len*2)
-		if fwopenresphdr.Status_Len != 0 {
-			err = items[1].DeSerialize(&extended_status)
+			newProps, err := client.GetAttrList(CipObject_ControllerInfo, 1, client.KeepAliveProps...)
 			if err != nil {
-				return fmt.Errorf("error deserializing forward open response header extended status. %w", err)
+				client.Logger.Printf("keepalive failed. %w", err)
+				client.SLogger.Error("keepalive failed", slog.Any("err", err))
+				client.Disconnect()
+				return
+			}
+			if newProps != originalProps {
+				client.Logger.Printf("controller change detected. re-analyzing types.\n Was: %+v\n  Is: %+v", originalProps, newProps)
+				client.SLogger.Info(
+					"controller change detected. re-analyzing types",
+					slog.Any("originalProps", originalProps),
+					slog.Any("newProps", newProps),
+				)
+				err := client.ListAllTags(0)
+				if err != nil {
+					client.Logger.Printf("keepalive list tags failed. %v", err)
+					return
+				}
+				originalProps = newProps
 			}
-		}
 
-		if fwopenresphdr.Status != 0x00 {
-			return fmt.Errorf("bad status on both forward opens. got %x", fwopenresphdr.Status)
+		case <-client.cancel_keepalive:
+			t.Stop()

```

Comment:
You are right. Should have been return instead of t.stop(). Not sure what I was doing there. I will update.

I believe the only time it should be stopped is if the client is disconnected or someone manually kills the keepalive.
If the client is disconnected the keepalive will come back on if KeppAlive == true.
If the keepalive is manually cancelled then I assume a user would know what they are doing and when to turn it back on if they want?